src/dawn/engine_dawn.cc

```diff
@@ -345,6 +343,49 @@ MapResult MapTextureToHostBuffer(const RenderPipelineInfo& render_pipeline,
   return map;
 }
 
+// Maps device buffers back to the host buffers
+Result MapDeviceBufferToHostBuffer(const ComputePipelineInfo& compute_pipeline,
+                                   const ::dawn::Device& device) {
+  for (uint32_t i = 0; i < compute_pipeline.pipeline->GetBuffers().size();
+       i++) {
+    auto& device_buffer = compute_pipeline.buffers[i];
+    auto& host_buffer = compute_pipeline.pipeline->GetBuffers()[i];
+
+    // Create a copy of device buffer to it in a map read operation.
+    // It's not possible to simply set this bit on the existing buffers since:
+    // Device error: Only CopyDst is allowed with MapRead
+    ::dawn::BufferDescriptor descriptor;
+    descriptor.size = host_buffer.buffer->GetSizeInBytes();
+    descriptor.usage =
+        ::dawn::BufferUsageBit::CopyDst | ::dawn::BufferUsageBit::MapRead;
+    const auto copy_device_buffer = device.CreateBuffer(&descriptor);
+    const uint64_t source_offset = 0;
+    const uint64_t destination_offset = 0;
+    const uint64_t copy_size =
+        static_cast<uint64_t>(host_buffer.buffer->GetSizeInBytes());
+    auto encoder = device.CreateCommandEncoder();
+    encoder.CopyBufferToBuffer(device_buffer, source_offset, copy_device_buffer,
+                               destination_offset, copy_size);
+    auto commands = encoder.Finish();
+    auto queue = device.CreateQueue();
+    queue.Submit(1, &commands);
+
+    MapResult mapped_device_buffer = MapBuffer(device, copy_device_buffer);
+    auto* values = host_buffer.buffer->ValuePtr();
+    values->resize(host_buffer.buffer->GetSizeInBytes());
+    // TODO(sarahM0): At the moment all compute tests pass. Investigate
+    // if ever alignment is an issue and if it's necessary to copy line by line
```

Comment:
- won't need to copy line by line because buffers are a big linear sequence of bytes. There's no enforced stride because there is no intrinsic "line".
- memcpy should work with byte-aligned inputs or outputs.

Reply:
ok makes sense